src/objects/capsule.rs

```diff
@@ -0,0 +1,548 @@
+                });
+                match CAPS_DATA {
+                    Some(Ok(d)) => Ok(d),
+                    Some(Err(ref e)) => Err(e.clone_ref(py)),
+                    _ => panic!("Uninitialized"), // can't happen
+                }
+            }
+        }
+    )
+}
+
+/// Macro to retrieve a function pointer capsule
+///
+/// For general explanations about Capsules, see [`PyCapsule`].
+///
+/// This macro takes the following arguments:
+/// - segments of the full Python dotted name of the capsule
+/// - `$rustmod`, a suitable name to define a module
+/// - signature of the wished function (the macro will insert the `extern unsafe fn`)
+///
+/// The macro defines a Rust module with a retrieval function such as:
+///
+/// ```ignore
+/// mod $rustmod {
+///     pub type CapsFn = unsafe extern "C" ( ... ) -> ... ;
+///     pub unsafe fn retrieve<'a>(py: Python) -> PyResult<CapsFn) { ... }
+/// }
+/// ```
+///
+/// # Examples
+/// There is in the Python library no capsule enclosing a function pointer directly,
+/// although the documentation presents it as a valid use-case. For this example, we'll
+/// therefore have to create one, using the [`PyCapsule`] constructor, and to set it in an
+/// existing  module (not to imply that a real extension should follow that example
+/// and set capsules in modules they don't define!)
+///
+///
+/// ```
+/// #[macro_use] extern crate cpython;
+/// extern crate libc;
+/// use cpython::{PyCapsule, Python, FromPyObject};
+/// use libc::{c_int, c_void};
+///
+/// extern "C" fn inc(a: c_int) -> c_int {
+///     a + 1
+/// }
+///
+/// /// for testing purposes, stores a capsule named `sys.capsfn`` pointing to `inc()`.
+/// fn create_capsule() {
+///     let gil = Python::acquire_gil();
+///     let py = gil.python();
+///     let pymod = py.import("sys").unwrap();
+///     let caps = PyCapsule::new(py, inc as *mut c_void, "sys.capsfn").unwrap();
+///     pymod.add(py, "capsfn", caps).unwrap();
+///  }
+///
+/// py_capsule_fn!(sys, capsfn, capsmod, (a: c_int) -> c_int);
+///
+/// // One could, e.g., reexport if needed:
+/// pub use capsmod::CapsFn;
+///
+/// fn retrieve_use_capsule() {
+///     let gil = Python::acquire_gil();
+///     let py = gil.python();
+///     let fun = capsmod::retrieve(py).unwrap();
+///     assert_eq!( unsafe { fun(1) }, 2);
+///
+///     // let's demonstrate the (reexported) function type
+///     let g: CapsFn = fun;
+/// }
+///
+/// fn main() {
+///     create_capsule();
+///     retrieve_use_capsule();
+///     retrieve_use_capsule();
+/// }
+/// ```
+/// [`PyCapsule`]: struct.PyCapsule.html
+#[macro_export]
+macro_rules! py_capsule_fn {
+    ($($capsmod:ident).+, $capsname:ident, $rustmod:ident, $( $sig: tt)* ) => (
+        mod $rustmod {
+            use super::*;
+            use std::sync::Once;
+            use $crate::PyClone;
+
+            pub type CapsFn = unsafe extern "C" fn $( $sig )*;
```

Comment:
No need to abbreviate `Capsule` here.  Calling this `CapsuleFn` would be clearer, I think.

Reply:
especially since it's `pub`, indeed. Done